File: python-gdb-activities-20/activity17/gdb/tests/test_command_logging.py
# gdb/tests/test_command_logging.py
import unittest
import os
from gdb.domain.account_factory import AccountFactory
from gdb.command.deposit_command import DepositCommand
from gdb.command.withdraw_command import WithdrawCommand
from gdb.command.transfer_command import TransferCommand
from gdb.service.transfer_service import TransferService
from gdb.logging.transaction_log import TransactionLog

class TestCommandLogging(unittest.TestCase):
    def setUp(self):
        # ============================================================
        # 📝 STEP 15: Create The Log, Accounts And Service
        #
        # INSTRUCTIONS:
        #   1. self.log_file = "test_transactions.log"
        #   2. self.log = TransactionLog(self.log_file), then call self.log.clear() so every test starts empty.
        #   3. self.acc1 = AccountFactory.create_account("SAVINGS", 1001, "Rajesh", 30, 50000.0) with PIN 1234.
        #   4. self.acc2 = AccountFactory.create_account("CURRENT", 1002, "Priya", 28, 50000.0) with PIN 5678.
        #   5. self.transfer_svc = TransferService()
        # ============================================================
        # TODO: create the transaction log, the two accounts and the transfer service
        raise NotImplementedError("TODO: Step 15 - create the log, accounts and transfer service")

    def tearDown(self):
        # Provided: delete the test log file after every test.
        self.log.clear()

    def test_deposit_command(self):
        # ============================================================
        # 📝 STEP 16: Execute And Log A DepositCommand
        #
        # INSTRUCTIONS:
        #   1. Create DepositCommand(self.acc1, 5000.0) and call execute().
        #   2. Assert get_transaction() is not None.
        #   3. Log it with self.log.log_transaction(...) and assert the log now holds exactly 1 transaction.
        # ============================================================
        # TODO: execute a deposit command, log its transaction and assert the log size
        raise NotImplementedError("TODO: Step 16 - test DepositCommand")

    def test_withdraw_command(self):
        # ============================================================
        # 📝 STEP 17: Execute And Log A WithdrawCommand
        #
        # INSTRUCTIONS:
        #   1. Create WithdrawCommand(self.acc1, 10000.0, 1234) and call execute().
        #   2. Assert get_transaction() is not None.
        #   3. Log it and assert the log now holds exactly 1 transaction.
        # ============================================================
        # TODO: execute a withdraw command, log its transaction and assert the log size
        raise NotImplementedError("TODO: Step 17 - test WithdrawCommand")

    def test_transfer_command(self):
        # ============================================================
        # 📝 STEP 18: Execute And Log A TransferCommand
        #
        # INSTRUCTIONS:
        #   1. Create TransferCommand(self.transfer_svc, self.acc1, self.acc2, 5000.0, 1234) and call execute().
        #   2. Assert get_transaction() is not None.
        #   3. Log it and assert the log now holds exactly 1 transaction.
        # ============================================================
        # TODO: execute a transfer command, log its transaction and assert the log size
        raise NotImplementedError("TODO: Step 18 - test TransferCommand")

    def test_file_persistence(self):
        # ============================================================
        # 📝 STEP 19: Verify File Persistence
        #
        # INSTRUCTIONS:
        #   1. Execute a DepositCommand(self.acc1, 3000.0) and log its transaction.
        #   2. Create a brand-new TransactionLog(self.log_file) and call load_from_file().
        #   3. Assert the new log loaded exactly 1 transaction from the file.
        #
        # HINT: The new instance starts empty, so the record can only come from the file on disk.
        # ============================================================
        # TODO: prove the logged transaction survives in the file
        raise NotImplementedError("TODO: Step 19 - test file persistence")

if __name__ == "__main__":
    unittest.main()

File: java-gdb-activities-20/activity17-solution/src/com/gdb/tests/TestCommandLogging.java
package com.gdb.tests;

import com.gdb.command.*;
import com.gdb.domain.*;
import com.gdb.logging.TransactionLog;
import java.util.List;

public class TestCommandLogging {
    public static void main(String[] args) throws Exception {
        System.out.println("=".repeat(60));
        System.out.println("  ACTIVITY 17 — COMMAND PATTERN + FILE LOGGING");
        System.out.println("=".repeat(60));

        TransactionLog log = new TransactionLog();
        log.clear(); // start fresh

        // 📝 STEP 10 — Create test accounts
        IAccount acc1 = AccountFactory.createAccount("SAVINGS", 1001, "John", 25, 15000);
        acc1.setPin(1234);
        IAccount acc2 = AccountFactory.createAccount("SAVINGS", 1002, "Jane", 30, 10000);

        // 📝 STEP 11 — Execute and log a DepositCommand
        DepositCommand depCmd = new DepositCommand(acc1, 5000);
        depCmd.execute();
        log.log(depCmd);
        System.out.println("[STEP 11] Logged: " + depCmd.getTransaction());

        // 📝 STEP 12 — Execute and log a WithdrawCommand
        WithdrawCommand wthCmd = new WithdrawCommand(acc1, 2000, 1234);
        wthCmd.execute();
        log.log(wthCmd);
        System.out.println("[STEP 12] Logged: " + wthCmd.getTransaction());

        // 📝 STEP 13 — Execute and log a TransferCommand
        TransferCommand trfCmd = new TransferCommand(acc1, acc2, 3000, 1234);
        trfCmd.execute();
        log.log(trfCmd);
        System.out.println("[STEP 13] Logged: " + trfCmd.getTransaction());

        // 📝 STEP 14 — Read all commands back
        List<TransactionCommand> history = log.readAll();
        System.out.println("\n[STEP 14] Read " + history.size() + " commands from transaction log:");
        for (int i = 0; i < history.size(); i++) {
            System.out.println("  [" + (i + 1) + "] " + history.get(i).getTransaction());
        }

        // 📝 STEP 15 — Verify persistence
        TransactionLog freshLog = new TransactionLog();
        List<TransactionCommand> persisted = freshLog.readAll();
        System.out.println("\n[STEP 15] Fresh reader verified " + persisted.size() + " persisted commands [PASS]");
    }
}
